doc/man/lfe.1
.\" Automatically generated by Pandoc 3.1.11
.\"
.TH "lfe" "1" "2008\-2020" "" ""
.SH NAME
lfe \- Lisp Flavoured Erlang (LFE) shell
.SH SYNOPSIS
\f[CR]lfe\f[R] is a simple LFE repl (read\-eval\-print loop) in which
you can enter sexprs which then are evaluated and the value printed.
You can also define local functions and macros as well as set variables.
It can read commands either from the standard input or from a file.
.PP
The LFE repl is implemented in the module \f[CR]lfe_shell\f[R].
.SH BUILT\-IN SHELL FUNCTIONS
These are defined as normal functions and macros and can be called from
anywhere in the shell.
They can even be redefined.
They can also be explicitly called (: lfe_shell \&...).
.PP
\f[B](c File [Options])\f[R]
.PP
Compile and load an LFE file.
Assumes default extension \f[CR].lfe\f[R].
.PP
\f[B](: c Command Arg \&...)\f[R]
.PP
\f[B](c:Command Arg \&...)\f[R]
.PP
All the commands in the standard Erlang shell can be reached in this
way.
.PP
\f[B](cd Dir)\f[R]
.PP
Change the working directory.
.PP
\f[B](clear)\f[R]
.PP
Clear the REPL output.
.PP
\f[B](ec File [Options])\f[R]
.PP
Compile and load an Erlang file.
.PP
\f[B](ep Expr [Depth])\f[R]
.PP
\f[B](epp Expr [Depth])\f[R]
.PP
Print/prettyprint a value in Erlang form to either the specified depth
or if no value is given the full depth.
.PP
\f[B](flush)\f[R]
.PP
Flush any messages sent to the shell.
.PP
\f[B](h)\f[R]
.PP
\f[B](help)\f[R]
.PP
Print usage info.
.PP
\f[B](h Mod)\f[R]
.PP
\f[B](h Mod Mac)\f[R]
.PP
\f[B](h Mod Fun Arity)\f[R]
.PP
Print out help information of a module/macro/function.
.PP
\f[B](i [(list Pid \&...)])\f[R]
.PP
Print information about a list of pids.
If no list is given then print information about currently running
processes in the system.
.PP
\f[B](i x y z)\f[R]
.PP
Print information about the about #Pid<x.y.z>
.PP
\f[B](l Module \&...)\f[R]
.PP
Load modules.
.PP
\f[B](ls)\f[R]
.PP
\f[B](ls dir)\f[R]
.PP
List files in a directory.
If no directory is given then list files in the current directory.
.PP
\f[B](m [Module \&...])\f[R]
.PP
Print out module information.
If no modules are given then print information about all modules.
.PP
\f[B](memory)\f[R]
.PP
**(memory type)
.PP
Returns the memory allocation information.
If a type or a list of types is given then only for those types.
.PP
\f[B](p Expr [Depth])\f[R]
.PP
\f[B](pp Expr [Depth])\f[R]
.PP
Print/prettyprint a value to either the specified depth or if no value
is given the full depth.
.PP
\f[B](pid x y z)\f[R]
.PP
Create a pid from x, y, z.
.PP
\f[B](pwd)\f[R]
.PP
Print the current working directory.
.PP
\f[B](q)\f[R]
.PP
Quit \- shorthand for \f[CR]init:stop/0\f[R].
.PP
\f[B](regs)\f[R]
.PP
Print information about the registered processes in the system.
.PP
\f[B](nregs)\f[R]
.PP
Print information about all the registered processes in the system.
.PP
\f[B](uptime)\f[R]
.PP
Print the node uptime.
.SH BUILT\-IN SHELL COMMANDS
These are special forms which are only recognised at the top\-level in
shell input.
The cannot be redefined.
.PP
\f[B](reset\-environment)\f[R]
.PP
Resets the environment to its initial state.
This will clear all variables, functions and macros that have been set.
.PP
\f[B](run File)\f[R]
.PP
Execute all the shell commands in File.
All defined variables, functions and macros will be saved in the
environment if there are no errors.
.PP
\f[B](set Pattern Expr)\f[R]
.PP
\f[B](set Pattern (when Guard) Expr)\f[R]
.PP
Evaluate Expr and match the result with Pattern binding variables in it.
These variables can then be used in the shell and also rebound in
another set.
.PP
\f[B](slurp File)\f[R]
.PP
Slurp in a source LFE file and makes all functions and macros defined in
the file available in the shell.
Only one file can be slurped at a time and slurping a new file basically
does an unslurp first.
.PP
\f[B](unslurp)\f[R]
.PP
Revert back to the state before the last slurp removing all function and
macro definitions both in the slurped file and defined in the shell
since then.
.SH SHELL FUNCTIONS AND MACROS
Functions and macros can be defined in the shell.
These will only be local to the shell and cannot be called from modules.
The forms are the standard forms for defining functions and macros.
.PP
\f[B](defun Fun \&...)\f[R]
.PP
Define a function in the shell.
.PP
\f[B](defmacro Macro \&...)\f[R]
.PP
Define a macro in the shell.
.SH BUILT\-IN SHELL VARIABLES
\f[B]\f[CB]+\f[B], \f[CB]++\f[B], \f[CB]+++\f[B]\f[R]
.PP
The three previous expressions input.
.PP
\f[B]\f[CB]*\f[B], \f[CB]**\f[B], \f[CB]***\f[B]\f[R]
.PP
The values of the previous three expressions.
.PP
\f[B]\f[CB]\-\f[B]\f[R]
.PP
The current expression input.
.SH SHELL ENVIRONMENT
The shell maintains an environment of local function and macro
definitions, and variable bindings.
The environment can be accessed using the built\-in shell variable $ENV.
This can be useful when calling functions like macroexpand and
macro\-function which unless an explicit environment is given will only
search the default environment.
.SH STARTING THE LFE SHELL
After installing the best way is probably to start Erlang directly
running the LFE shell with:
.IP
.EX
lfe [flags]
.EE
.PP
From a normal Erlang shell the best way to start the shell is by
calling:
.IP
.EX
17> lfe_shell:server().
.EE
.PP
Giving the user switch commands:
.IP
.EX
\-\-> s lfe_shell
\-\-> c
.EE
.PP
will create a job running the LFE shell and connect to it.
This also works when starting a remote shell.
.PP
Flags that LFE recognizes include the following:
.IP \[bu] 2
\f[CR]\-nobanner\f[R] \- starts LFE without showing the banner
.IP \[bu] 2
\f[CR]\-h\f[R] or \f[CR]\-\-help\f[R] \- provides command line usage
help
.IP \[bu] 2
\f[CR]\-e\f[R] or \f[CR]\-eval\f[R] \- evaluates a given sexpr in a
string
.IP \[bu] 2
\f[CR]\-prompt\f[R] \- users may supply a value here to override the
default \f[CR]lfe>\f[R] prompt; note that \f[CR]\-prompt classic\f[R]
will set the prompt to the original \f[CR]>\f[R] and \f[CR]\-prompt\f[R]
with no associated value will cause no prompt to be displayed at all.
These also work when node names are provided (with either
\f[CR]\-sname\f[R] or \f[CR]\-name\f[R]).
Furthermore, users may override the default formatting of node names in
prompts by providing a prompt value containing the string
\f[CR]\[ti]node\f[R] (which will be substituted with the actual name of
the node).
.PP
There can be multiple string expressions to be evaluated; each one must
be prefixed with an \f[CR]\-e\f[R] or \f[CR]\-eval\f[R].
String expressions are run in the LFE repl so shell commands and
functions are allowed.
They are all run in the same invocation of the repl so:
.IP
.EX
$ lfe \-e \[dq](set aaa 42)\[dq] \-e \[dq](set bbb 84)\[dq] \-e \[dq](pp (tuple aaa bbb))\[dq]
#(42 84)
.EE
.PP
If there are string expressions then the LFE repl will \f[CR]not\f[R] be
run.
.SH RUNNING LFE SHELL SCRIPTS
The LFE shell can also be directly called to run LFE shell scripts with:
.IP
.EX
lfe [flags] file [args]
.EE
.PP
This will start the shell, run a script with LFE shell commands and then
terminate the shell.
The following built\-in variables are also bound:
.PP
\f[B]script\-name\f[R]
.PP
The name of the script file as a string.
.PP
\f[B]script\-args\f[R]
.PP
A list of the arguments to the script as strings.
If no arguments have been given then this will be an empty list.
.PP
Note that if there are any string expressions to be evaluated then these
must come before the name of the script file and its arguments.
These expressions will be evaluated before the script and the script
will use the environment from the string expressions.
.PP
It is possible to run both string expressions and an LFE shell script
and they are then run in the same LFE repl.
.SH SEE ALSO
\f[B]lfescript(1)\f[R], \f[B]lfe_guide(7)\f[R] \f[B]lfe_doc(3)\f[R]
.SH AUTHORS
Robert Virding.


src/ChangeLog
2013-02-12  Robert Virding  <rv@renat.local>

    * <All files>: Changed Copyright notice and now use Apache License
    V2.0.

2011-08-19  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (pass_expand_expr): Only try to expand lists.

2011-07-09  Robert Virding  <rv@renat.local>

    * lfe_trans.erl (from_bitseg): Fix handling of binaries.

2011-07-02  Robert Virding  <rv@renat.local>

    * lfe_eval.erl (match_binary): Fix handling of binary and pattern
    vars.

    * lfe_lint.erl (pat_binary): Fix handling of binary and pattern
    vars.

2011-06-29  Robert Virding  <rv@renat.local>

    * lfe_lint.erl, lfe_codegen.erl, lfe_eval.erl, lfe_trans.erl: Now
    use lfe_bits to parse binaries.

2011-06-26  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Add match-spec and qlc macros.

2011-06-15  Robert Virding  <rv@renat.local>

    * lfe_shell.erl (macroexpand, macroexpand-1, macroexpand-all): No
    longer shell functions.

    * lfe_lib.erl (del_vbinding): New function.

2011-06-14  Robert Virding  <rv@renat.local>

    * lfe_io_pretty.erl (indent_type): Add funcall.

2011-06-08  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Add match-spec macro.

2011-06-05  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Add simple include-lib macro.

2011-05-30  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_defrecord): Add ets/mnesia match pattern
    macro emp-XXX.

    * lfe_eval.erl, lfe_shell.erl, lfe_macro.erl: Restructure
    lfe_macro.erl and redefine interface.

2011-04-24  Robert Virding  <rv@renat.local>

    * lfe_codegen.erl (comp_gtest): Removed unnecessary special cases,
    optimizer fixes this for us.

2011-04-21  Robert Virding  <rv@renat.local>

    * lfe_lint.erl: Remove unused erroneous lfe_lib imports.

2011-04-19  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Arithmetic and comparison functions
    take multiple arguments.

2011-04-12  Robert Virding  <rv@renat.local>

    * lfe_comp.erl (erl_comp_opts, do_ok_return): Handle
    warnings_as_errors ourselves in same way as vanilla compiler.

2011-04-08  Robert Virding  <rv@renat.local>

    * lfe_comp.erl (when_opt): Change argument order.

2011-03-16  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (expand_append): Expand ++ in such a way as to
    allow use in patterns.
    (expand_defmacro): Allow (defmacro name arg ...).

2010-12-31  Robert Virding  <rv@renat.local>

    * lfe_io_format.erl (control, newline): Handle ~n correctly.

    * lfe_io_pretty.erl (print1): Prettyprint one element lists.

2010-11-25  Robert Virding  <rv@renat>

    * lfe_lint.erl (pat_bitsegs): Use check_foldl instead.
    (gexpr_bitsegs): Use check_foreach instead.

2010-11-04  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Extend ? macro to handle optional
    timeout and default value.

2010-11-01  Robert Virding  <rv@renat>

    * lfe_scan.xrl: Check creating atoms/symbols and code cleanup.

2010-10-27  Robert Virding  <rv@renat.local>

    * lfe_shell.erl (prompt): Prompt adds node name when alive.

2010-09-06  Robert Virding  <rv@stanislaw.local>

    * lfe_macro.erl (def_record, def_rec_fields): Record macros follow
    new pattern code.
    (expand_defmacro): Add explicit list constructor for args.
    (c_l_tq): Add explicit cons constructor.

    * lfe_codegen.erl (comp_pat, pat_list): Add code to generate
    pattern code for explicit list constructors.

    * lfe_eval.erl (match, match_list): Add code to match explicit
    list constructors.

    * lfe_lint.erl (check_pat, pat_list, check_alias): Add code to
    test for explicit list constructors.

2010-08-25  Robert Virding  <rv@stanislaw.local>

    * lfe_macro.erl (c_tq, c_l_tq, c_b_tq): Add guards to generators
    in comprehensions.

    * lfe_lint.erl (format_error, check_gbody, check_gexpr)
    (gexpr_bitsegs): Better error checking in guards.

2010-07-24  Robert Virding  <rv@host-90-237-128-188.mobileonline.telia.com>

    * lfe_shell.erl (macroexpand, macroexpand_1): Return evaluated
    argument if no expansion.

2010-07-12  Robert Virding  <rv@host-90-233-145-108.mobileonline.telia.com>

    * lfe_scan.xrl: Have separate cases for integer, float and symbol
    atoms, catch errors better.

2010-06-07  Robert Virding  <rv@stanislaw.local>

    * lfe_lint.erl (expr_bitseg, gexpr_bitseg, pat_bitseg): Allow
    strings (integer lists).

    * lfe_codegen.erl (parse_bitsegs): Allow strings (integer lists).

    * lfe_eval.erl (parse_bitsegs): Allow strings (integer lists).

    * lfe_eval.lfe (eval-gbinary): Evaluate binaries in guards.

    * lfe_eval.erl (eval_gbinary): Evaluate binaries in guards.

2010-06-03  Robert Virding  <rv@stanislaw.local>

    * All files: Update copyright to 2010.

2010-04-30  Robert Virding  <rv@stanislaw.local>

    * lfe_eval.lfe (apply-lambda): New name and more logical argument
    order for eval-lambda/4.
    (apply-match-clauses): New name and more logical argument
    order for eval-match-clauses/3.

2010-02-24  Robert Virding  <rv@stanislaw.local>

    * lfe_eval.erl (apply_lambda): New name and more logical argument
    order for eval_lambda/4.
    (apply_match_clauses): New name and more logical argument
    order for eval_match_clauses/3.

    * lfe_macro.erl (expand_clause, expand_ml_clause, exp_predef)
    (c_tq): Guards are now a sequence of tests (when t ...).

2010-02-20  Robert Virding  <rv@stanislaw.local>

    * lfe_io.erl (scan_and_parse): Use new re-entrant parser.

    * lfe_parse.erl: Write to become re-entrant using LL(1) grammar.

2010-02-16  Robert Virding  <rv@stanislaw.local>

    * lfe_eval.lfe (eval-guard, match-when, eval-let): Guards are now
    a sequence of tests (when t ...).

    * lfe_eval.erl (eval_guard, match_when, eval_let): Guards are now
    a sequence of tests (when t ...).

    * lfe_codegen.erl (comp_guard, comp_gtest, comp_clause_body)

src/lfe.erl
%% Copyright (c) 2016-2020 Robert Virding
%%
%% Licensed under the Apache License, Version 2.0 (the "License");
%% you may not use this file except in compliance with the License.
%% You may obtain a copy of the License at
%%
%%     http://www.apache.org/licenses/LICENSE-2.0
%%
%% Unless required by applicable law or agreed to in writing, software
%% distributed under the License is distributed on an "AS IS" BASIS,
%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
%% See the License for the specific language governing permissions and
%% limitations under the License.

%% File    : lfe.erl
%% Author  : Robert Virding
%% Purpose : Lisp Flavoured Erlang standard library.

-module(lfe).

%% Standard lisp library.
-export([acons/3,pairlis/2,pairlis/3,
         assoc/2,'assoc-if'/2,'assoc-if-not'/2,
         rassoc/2,'rassoc-if'/2,'rassoc-if-not'/2,
         subst/3,'subst-if'/3,'subst-if-not'/3,sublis/2,
         eval/1,eval/2,
         'macro-function'/1,'macro-function'/2,
         macroexpand/1,macroexpand/2,
         'macroexpand-1'/1,'macroexpand-1'/2,
         'macroexpand-all'/1,'macroexpand-all'/2]).

-export(['LFE-EXPAND-EXPORTED-MACRO'/3]).

%% Metadata
-export([version/0]).

%% 'LFE-EXPAND-EXPORTED-MACRO'(Name, Args, Env) -> {yes,Expansion} | no.
%%  A smart implementation of this where we call the lfe_macro module
%%  to expand and check for us. Using an empty environment ensures
%%  only predefined macros are used.

'LFE-EXPAND-EXPORTED-MACRO'(Name, Args, _) ->
    lfe_macro:expand_expr([Name|Args], lfe_env:new()).

%% Standard lisp library functions.

%% Association list functions.

acons(K, V, Alist) -> [[K|V]|Alist].

pairlis(Ks, Vs) -> pairlis(Ks, Vs, []).

pairlis([K|Ks], [V|Vs], Alist) ->
    [[K|V]|pairlis(Ks, Vs, Alist)];
pairlis([], [], Alist) -> Alist.

assoc(K, [[K|_]=Pair|_]) -> Pair;
assoc(K, [_|L]) -> assoc(K, L);
assoc(_, []) -> [].

'assoc-if'(Pred, [[K|_]=Pair|L]) ->
    case Pred(K) of
        true -> Pair;
        false -> 'assoc-if'(Pred, L)
    end;
'assoc-if'(_, []) -> [].

'assoc-if-not'(Pred, [[K|_]=Pair|L]) ->
    case Pred(K) of
        false -> Pair;
        true -> 'assoc-if-not'(Pred, L)
    end;
'assoc-if-not'(_, []) -> [].

rassoc(V, [[_|V]=Pair|_]) -> Pair;
rassoc(V, [_|L]) -> rassoc(V, L);
rassoc(_, []) -> [].

'rassoc-if'(Pred, [[_|V]=Pair|L]) ->
    case Pred(V) of
        true -> Pair;
        false -> 'rassoc-if'(Pred, L)
    end;
'rassoc-if'(_, []) -> [].

'rassoc-if-not'(Pred, [[_|V]=Pair|L]) ->
    case Pred(V) of
        false -> Pair;
        true -> 'rassoc-if-not'(Pred, L)
    end;
'rassoc-if-not'(_, []) -> [].

%% subst(New, Old, Tree) -> Tree.

subst(New, Old, Old) -> New;
subst(New, Old, [H|T]) ->
    [subst(New, Old,H)|subst(New, Old, T)];
subst(_, _, Tree) -> Tree.

%% subst-if(New, Test, Tree) -> Tree.

'subst-if'(New, Test, Tree) ->
    case Test(Tree) of
        true -> New;
        false ->
            case Tree of
                [H|T] ->
                    ['subst-if'(New, Test, H)|'subst-if'(New, Test, T)];
                _ -> Tree
            end
    end.

%% subst-if-not(New, Test, Tree) -> Tree.

'subst-if-not'(New, Test, Tree) ->
    case Test(Tree) of
        false -> New;
        true ->
            case Tree of
                [H|T] ->
                    ['subst-if-not'(New, Test, H)|'subst-if-not'(New, Test, T)];
                _ -> Tree
            end
    end.

%% sublis(AList, Tree) -> Tree.

sublis(Alist, Tree) ->
    case assoc(Tree, Alist) of
        [_|New] -> New;                         %Found it
        [] ->                                   %Not there
            case Tree of
                [H|T] ->
                    [sublis(Alist, H)|sublis(Alist, T)];
                _ -> Tree
            end
    end.

%%  eval(Sexpr) -> Value.
%%  macro-function(Name [,Environment]) -> Macro | [].
%%  macroexpand(Form [,Environment]) -> Expansion | Form.
%%  macroexpand-1(Form [,Environment]) -> Expansion | Form.
%%  macroexpand-all(Form [,Environment]) -> Expansion | Form.

eval(Sexpr) -> eval(Sexpr, lfe_env:new()).      %Empty environment.
eval(Sexpr, Env) -> lfe_eval:expr(Sexpr, Env).

'macro-function'(Symb) -> 'macro-function'(Symb, lfe_env:new()).
'macro-function'(Symb, Env) ->
    case lfe_env:get_mbinding(Symb, Env) of
        {yes,Macro} ->
            Macro;
        no -> []
    end.

macroexpand(Form) -> macroexpand(Form, lfe_env:new()).
macroexpand(Form, Env) ->
    case lfe_macro:expand_expr(Form, Env) of
        {yes,Exp} -> Exp;
        no -> Form
    end.

'macroexpand-1'(Form) -> 'macroexpand-1'(Form, lfe_env:new()).
'macroexpand-1'(Form, Env) ->
    case lfe_macro:expand_expr_1(Form, Env) of
        {yes,Exp} -> Exp;
        no -> Form
    end.

'macroexpand-all'(Form) -> 'macroexpand-all'(Form, lfe_env:new()).
'macroexpand-all'(Form, Env) -> lfe_macro:expand_expr_all(Form, Env).

%% Metadata

version() ->
    {ok, [App]} = file:consult(code:where_is_file("lfe.app")),
    proplists:get_value(vsn, element(3, App)).
